zebra-state/src/service.rs

```diff
@@ -149,9 +149,14 @@ impl StateService {
         let parent_hash = prepared.block.header.previous_block_hash;
 
         if self.disk.finalized_tip_hash() == parent_hash {
-            self.mem.commit_new_chain(prepared);
+            self.mem
+                .commit_new_chain(prepared, self.disk.mmr().clone())?;
         } else {
-            self.mem.commit_block(prepared);
+            let block_iter = self.any_ancestor_blocks(parent_hash);
+            // TODO: the above creates a immutable borrow of self, and below it uses
+            // a mutable borrow. which causes a compiler error. How to solve this?
+            self.mem
+                .commit_block(prepared, self.disk.mmr(), &block_iter)?;
```

Comment:
`commit_block` needs the following info to create an up-to-date MMR:
- the finalized tip MMR
- the blocks from the finalized tip child to the fork block
- the new block we're about to add

The chain that `commit_block` clones already contains all the blocks from the finalized tip child to the fork block. So once you've done the rollback to the fork block, `Chain.blocks` should contain exactly the blocks you need.

Reply:
Thanks! For some reason I hadn't considered using `Chain.blocks` and was assuming I had to get the blocks from the Service or NonFinalizedState. Now the code is much simpler.